org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelCompletionParticipant.java

```diff
@@ -302,30 +303,68 @@ public void onAttributeName(boolean generateValue, ICompletionRequest request, I
 		if (attributes == null) {
 			return;
 		}
 		for (CMAttributeDeclaration attributeDeclaration : attributes) {
-			if (parentElement != null) {
-				String prefix = parentElement.getPrefix(attributeDeclaration.getNamespace());
-				String attrName = attributeDeclaration.getName(prefix);
-				if (!parentElement.hasAttribute(attrName)) {
-					CompletionItem item = new AttributeCompletionItem(attrName, canSupportSnippet, fullRange,
-							generateValue,
-							attributeDeclaration.getDefaultValue(), attributeDeclaration.getEnumerationValues(),
-							request.getSharedSettings());
-					if (request.isResolveDocumentationSupported()) {
-						addResolveData(request, item, AttributeNameCompletionResolver.PARTICIPANT_ID);
-					} else {
-						MarkupContent documentation = XMLGenerator.createMarkupContent(attributeDeclaration,
-								elementDeclaration,
-								request);
-						item.setDocumentation(documentation);
+			String prefix = null;
+			boolean declareNamespace = false;
+			String namespaceURI = attributeDeclaration.getNamespace();
+			if (!StringUtils.isEmpty(namespaceURI)) {
+				// The schema defines the attribute with a namespace (ex :
+				// http://www.w3.org/1999/xlink)
+				// Try to get the prefix from the XML document (ex: <docbook
+				// xlink="http://www.w3.org/1999/xlink")
+				prefix = parentElement.getPrefix(namespaceURI);
+				if (prefix == null) {
+					// The namespace is not declared in the XML document, get the prefered prefeix
+					// defined in the grammar.
+					prefix = attributeDeclaration.getPrefix();
+					// The xlink="http://www.w3.org/1999/xlink" needs to be inserted in the root
+					// document element.
+					declareNamespace = true;
+				}
+			}
+			String attrName = attributeDeclaration.getName(prefix);
+			if (!parentElement.hasAttribute(attrName)) {
+				CompletionItem item = new AttributeCompletionItem(attrName, canSupportSnippet, fullRange,
+						generateValue,
+						attributeDeclaration.getDefaultValue(), attributeDeclaration.getEnumerationValues(),
+						request.getSharedSettings());
+				if (declareNamespace) {
+					// Insert with additional text edit, the declaration of the namespace (ex :
+					// xlink="http://www.w3.org/1999/xlink")
+					DOMDocument document = parentElement.getOwnerDocument();
+					int offset = document.getDocumentElement().getOffsetBeforeCloseOfStartTag();
+					Range range = XMLPositionUtility.createRange(offset, offset, document);
+
+					StringBuilder insertText = new StringBuilder();
+					insertText.append(' ');
+					insertText.append("xmlns:");
+					insertText.append(prefix);
+					insertText.append("=\"");
+					insertText.append(namespaceURI);
+					insertText.append("\"");
+					char currentChar = document.getText().charAt(offset);
+					char previousChar = document.getText().charAt(offset - 1);
+					if (currentChar == '>' && previousChar == ' ') {
+						insertText.append(' ');

```

Comment:
I think in this case, we don't want to insert the space inserted on line 346 since it ends up being 2 spaces: 
![Peek 2023-02-22 11-20](https://user-images.githubusercontent.com/73968480/220689043-86f6cebd-8212-434a-9fae-fc3ed0d6c163.gif)

Reply:
Good catch. It should be fixed now (with tests too).